category_theory/discrete_category.lean

```diff
@@ -0,0 +1,77 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Stephen Morgan, Scott Morrison
+
+import data.ulift
+import category_theory.natural_transformation
+import category_theory.isomorphism
+import category_theory.functor_category
+
+namespace category_theory
+
+universes u₁ v₁ u₂ v₂
+
+def discrete (α : Type u₁) := α
+
+@[extensionality] lemma plift.ext {P : Prop} (a b : plift P) : a = b :=
+begin
+  cases a, cases b, refl
+end
+
+instance plift_subsingleton (P : Prop) : subsingleton (plift P) :=
+subsingleton.intro (λ a b, begin cases a, cases b, refl end)
+
+instance discrete_category (α : Type u₁) : small_category (discrete α) :=
+{ hom  := λ X Y, ulift (plift (X = Y)),
+  id   := by obviously,
+  comp := by obviously }
+
+-- TODO this needs to wait for equivalences to arrive
+-- example : equivalence.{u₁ u₁ u₁ u₁} punit (discrete punit) := by obviously
+
+def discrete.lift {α : Type u₁} {β : Type u₂} (f : α → β) : (discrete α) ⥤ (discrete β) :=
+{ obj := f,
+  map := λ X Y g, begin cases g, cases g, cases g, exact 𝟙 (f X) end }
+
+variables (J : Type v₂) [small_category J]
+
+variables (C : Type u₂) [𝒞 : category.{u₂ v₂} C]
+include 𝒞
+
+@[simp] def discrete.forget : discrete J ⥤ J :=
```

Comment:
`forget` is a bit of a funny name here. We could call it `discrete.counit` because it is the counit of the adjunction (discrete : Set -> Cat, ob : Cat -> Set). But do we really need it at all, I wonder?

Reply:
Removed, per above.